renku/core/commands/checks/storage.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Check for large files in Git history."""
+import re
+import shlex
+from subprocess import PIPE, STDOUT, run
+
+from renku.core import errors
+from renku.core.commands.echo import WARNING
+
+
+def check_lfs_info(client):
+    """Checks if files in history should be in LFS."""
+    if not client.has_external_storage:
+        return True, None
+
+    exclude_patterns = _read_lfs_exclude_patterns(client)
+    above_size = client.minimum_lfs_file_size
+
+    excludes = [
+        p for e in exclude_patterns for p in ('--exclude', shlex.quote(e))
+    ]
+
+    lfs_migrate_cmd = [
+        'git', 'lfs', 'migrate', 'info', '--include-ref',
+        client.repo.active_branch.name, '--above',
+        str(above_size), '--top', '42000'
+    ] + excludes
+
+    try:
+        lfs_output = run(
+            lfs_migrate_cmd,
+            stdout=PIPE,
+            stderr=STDOUT,
+            cwd=client.path,
+            universal_newlines=True
+        )
+    except (KeyboardInterrupt, OSError) as e:
+        raise errors.GitError(
+            'Couldn\'t run \'git lfs migrate info\':\n{0}'.format(e)
+        )
+
+    files = []
+    files_re = re.compile(r'(.*\s+[\d.]+\s+\S+).*')
+
+    for line in lfs_output.stdout.split('\n'):
+        match = files_re.match(line)
+        if match:
+            files.append(match.groups()[0])
+
+    if not files:
+        return True, None
+
+    message = (
+        WARNING + 'Git history contains files that should be in LFS.\n\t' +
+        '\n\t'.join(files) + '\n'
+    )
+
+    return False, message
+
+
+def _read_lfs_exclude_patterns(client):
+    lfs_ignore_path = client.path / client.RENKU_LFS_IGNORE_PATH
+
+    try:
+        lines = open(lfs_ignore_path).read().split('\n')
+    except FileNotFoundError:
+        return []
+
+    return [l for l in lines if l.strip() and not l.strip().startswith('#')]
```

Comment:
This won't work (There's also positive patterns like `!file_to_include`) which would need to be passed as `--include`.

I don't like us parsing the ignore file manually here, that's what https://github.com/SwissDataScienceCenter/renku-python/blob/master/renku/core/management/storage.py#L104 is meant for. It is unfortunate that the pathspec library compiles patterns to regex internally and doesn't store the original pattern.

Maybe instead we could override https://github.com/cpburnz/python-path-specification/blob/3aee306db792bbabaf41d31f99441f94b1b3e740/pathspec/pattern.py#L74 in renku/core/compat.py to also do `self.pattern = pattern` in __init__. Then we could still use `client.renku_lfs_ignore` here, access the `patterns` property to get all patterns, check `pattern.include` to see if it's an include/exclude/nop (though an include means exclude for our purposes and vice-versa) and use the new `pattern.pattern` to get the pattern. `pathspec.from_lines` does allow passing in an arbitrary class for pattern matching.

What do you think?

Reply:
Good catch and nice idea! I've noticed that pathspec does not save the original pattern but it did not pass my mind that we can register a custom matcher.